File: python/scripts/test_dependency_matrix.py
"""Test Action Guard across Python and dependency version ranges.

This script uses uv to provision Python interpreters, creates per-scenario virtual
environments, pins the dependency under test, resolves the remaining dependencies
within their supported ranges, installs the local package, and runs a small smoke
import check.
"""

from __future__ import annotations

import argparse
import os
import re
import shutil
import subprocess
from pathlib import Path

try:
    from packaging.version import InvalidVersion, Version
except ImportError:  # pragma: no cover
    InvalidVersion = ValueError
    Version = None


DEFAULT_PYTHON_VERSIONS = (
    "3.8",
    "3.9",
    "3.10",
    "3.11",
    "3.12",
    "3.13",
    "3.14",
)

CORE_RANGES = {
    "numpy": ("1.21.6", "3.0.0"),
    "onnxruntime": ("1.14.0", "2.0.0"),
    "openai": ("1.0.0", "4.0.0"),
}

OPTIONAL_RANGES = {
    "python-dotenv": ("0.21.0", "2.0.0"),
    "requests": ("2.28.0", "3.0.0"),
    "rich": ("12.0.0", "15.0.0"),
}

NUMPY_2_ONNXRUNTIME_MIN = "1.19.0"


def _run(cmd, cwd=None, env=None):
    result = subprocess.run(
        cmd,
        cwd=cwd,
        env=env,
        check=True,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
    )
    return result.stdout


def _uv_available():
    return shutil.which("uv") is not None


def _normalize_python_versions(values):
    versions = []
    for value in values or DEFAULT_PYTHON_VERSIONS:
        for version in re.split(r"[,\s]+", value.strip()):
            if not version:
                continue
            if not re.fullmatch(r"\d+\.\d+", version):
                raise ValueError(
                    f"Invalid Python version {version!r}; expected MAJOR.MINOR"
                )
            if version not in versions:
                versions.append(version)
    return versions


def _python_version_tuple(version):
    major, minor = version.split(".", maxsplit=1)
    return int(major), int(minor)


def _dependency_ranges_for_python(python_version):
    ranges = dict(CORE_RANGES)
    tokenizers_min = (
        "0.21.0" if _python_version_tuple(python_version) >= (3, 13) else "0.13.3"
    )
    ranges["tokenizers"] = (tokenizers_min, "1.0.0")
    return ranges


def _parse_versions(pip_output):
    match = re.search(r"Available versions:\s*(.*)", pip_output)
    if not match:
        return []
    raw = match.group(1)
    return [v.strip() for v in raw.split(",") if v.strip()]


def _filter_versions(versions, min_v, max_v):
    if Version is None:
        return versions

    min_ver = Version(min_v)
    max_ver = Version(max_v)
    filtered = []
    for v in versions:
        try:
            parsed = Version(v)
        except InvalidVersion:
            continue
        if min_ver <= parsed < max_ver:
            filtered.append(v)
    return filtered


def _sample_versions(versions, count):
    if not versions:
        return []
    if count <= 1:
        return [versions[0]]
    if len(versions) <= count:
        return versions

    step = max(1, len(versions) // (count - 1))
    sampled = [versions[0]]

    idx = step
    while len(sampled) < count - 1 and idx < len(versions) - 1:
        sampled.append(versions[idx])
        idx += step

    sampled.append(versions[-1])
    return sampled


def _version_is_installable(python, package, version):
    try:
        _run(
            [
                "uv",
                "pip",
                "install",
                "-p",
                python,
                "--dry-run",
                "--only-binary",
                ":all:",
                f"{package}=={version}",
            ]
        )
    except (OSError, subprocess.CalledProcessError):
        return False
    return True


def _resolve_versions(python, package, min_v, max_v, samples):
    try:
        output = _run([python, "-m", "pip", "index", "versions", package])
    except (OSError, subprocess.CalledProcessError) as exc:
        raise RuntimeError(f"Unable to query versions for {package}") from exc

    versions = _filter_versions(list(reversed(_parse_versions(output))), min_v, max_v)
    if not versions:
        raise RuntimeError(
            f"No published {package} versions found in range >={min_v},<{max_v}"
        )

    probe_cache = {}

    def installable(index):
        version = versions[index]
        if version not in probe_cache:
            probe_cache[version] = _version_is_installable(python, package, version)
        return probe_cache[version]

    target_count = min(samples, len(versions))
    if target_count == 1:
        target_indices = [0]
    else:
        target_indices = [
            round(index * (len(versions) - 1) / (target_count - 1))
            for index in range(target_count)
        ]

    selected = []
    for target_index in target_indices:
        for distance in range(len(versions)):
            candidate_indices = [target_index - distance]
            if distance:
                candidate_indices.append(target_index + distance)

            found = None
            for candidate_index in candidate_indices:
                if not 0 <= candidate_index < len(versions):
                    continue
                if installable(candidate_index):
                    found = versions[candidate_index]
                    break

            if found is not None:
                if found not in selected:
                    selected.append(found)
                break

    if len(selected) < target_count:
        for index, version in enumerate(versions):
            if version in selected or not installable(index):
                continue
            selected.append(version)
            if len(selected) >= target_count:
                break

    if not selected:
        raise RuntimeError(
            f"No installable {package} versions found for {python} in range "
            f">={min_v},<{max_v}"
        )

    if Version is not None:
        selected.sort(key=Version)

    return _sample_versions(selected, samples)


def _build_scenarios(
    core_versions, optional_versions, include_optional, bounds_only=False
):
    scenarios = []

    base_min = {name: versions[0] for name, versions in core_versions.items()}
    if include_optional:
        base_min.update(
            {name: versions[0] for name, versions in optional_versions.items()}
        )
    scenarios.append(("base-min", base_min))

    base_max = {name: versions[-1] for name, versions in core_versions.items()}
    if include_optional:
        base_max.update(
            {name: versions[-1] for name, versions in optional_versions.items()}
        )
    scenarios.append(("base-max", base_max))

    if bounds_only:
        return scenarios

    for dep_name, versions in core_versions.items():
        for v in versions:
            scenario = base_min.copy()
            scenario[dep_name] = v
            scenarios.append((f"core-{dep_name}-{v}", scenario))

    if include_optional:
        for dep_name, versions in optional_versions.items():
            for v in versions:
                scenario = base_min.copy()
                scenario[dep_name] = v
                scenarios.append((f"opt-{dep_name}-{v}", scenario))

    return scenarios


def _scenario_target_dependency(scenario_name, dependency_names):
    for dependency_name in sorted(dependency_names, key=len, reverse=True):
        for prefix in ("core-", "opt-"):
            if scenario_name.startswith(f"{prefix}{dependency_name}-"):
                return dependency_name
    return None


def _requirements_for_scenario(
    scenario_name,
    scenario_deps,
    core_ranges,
    optional_ranges,
):
    supported_ranges = dict(core_ranges)
    supported_ranges.update(optional_ranges)
    target_dependency = _scenario_target_dependency(scenario_name, scenario_deps.keys())
    target_version = scenario_deps.get(target_dependency)
    numpy_target_is_v2 = (
        target_dependency == "numpy"
        and target_version is not None
        and int(target_version.split(".", 1)[0]) >= 2
    )

    requirements = []
    for dependency_name, selected_version in scenario_deps.items():
        if dependency_name == target_dependency:
            requirements.append(f"{dependency_name}=={selected_version}")
            continue

        min_version, max_version = supported_ranges[dependency_name]
        if (
            dependency_name == "onnxruntime"
            and numpy_target_is_v2
            and (
                Version is None
                or Version(min_version) < Version(NUMPY_2_ONNXRUNTIME_MIN)
            )
        ):
            min_version = NUMPY_2_ONNXRUNTIME_MIN
        requirements.append(f"{dependency_name}>={min_version},<{max_version}")

    resolution = "highest" if scenario_name == "base-max" else "lowest-direct"
    return requirements, resolution


def _ensure_python(python_version):
    if not _uv_available():
        raise RuntimeError("uv is required to run the dependency matrix")

    print(f"Ensuring Python {python_version} is available with uv")
    _run(["uv", "python", "install", python_version])


def _create_venv(venv_path, python_version):
    if venv_path.exists():
        shutil.rmtree(venv_path)

    if not _uv_available():
        raise RuntimeError("uv is required to run the dependency matrix")

    print(f"Creating Python {python_version} venv with uv at {venv_path}")
    _run(["uv", "venv", "--python", python_version, str(venv_path)])


def _venv_python(venv_path):
    if os.name == "nt":
        return str(venv_path / "Scripts" / "python.exe")
    return str(venv_path / "bin" / "python")


def _assert_python_version(python, expected_version):
    version = _run(
        [
            python,
            "-c",
            "import sys; print(f'{sys.version_info.major}.{sys.version_info.minor}')",
        ]
    ).strip()

    if version != expected_version:
        raise RuntimeError(f"Expected Python {expected_version}, got {version}")


def _install_dependencies(python, requirements, resolution):
    print("Installing dependencies:")
    for requirement in requirements:
        print(f"  - {requirement}")
    print(f"Resolution strategy: {resolution}")

    print("Using uv to install dependencies")
    _run(["uv", "pip", "install", "-p", python, "-U", "pip", "setuptools", "wheel"])
    _run(
        [
            "uv",
            "pip",
            "install",
            "-p",
            python,
            "--only-binary",
            ":all:",
            "--resolution",
            resolution,
            *requirements,
        ]
    )


def _install_package(python, package_root):
    print("Installing local package (editable) with uv")
    _run(
        [
            "uv",
            "pip",
            "install",
            "-p",
            python,
            "-e",
            ".",
            "--no-deps",
        ],
        cwd=package_root,
    )


def _run_smoke_test(python, include_optional):
    env = os.environ.copy()
    env["EMBED_MODEL_NAME"] = "test-embedding"

    print("Running smoke test")

    snippet_lines = [
        "import agent_action_guard.action_classifier as ac",
        "import openai",
        "import tokenizers",
        "assert ac.classifier.session is not None",
        "openai_major = int(openai.__version__.split('.', 1)[0])",
        "client = openai.OpenAI(api_key='test-key') if openai_major >= 3 else None",
        "assert openai_major < 3 or client.embeddings is not None",
    ]

    if include_optional:
        snippet_lines += [
            "import dotenv",
            "import requests",
            "import rich",
        ]

    snippet = ";".join(snippet_lines)
    _run([python, "-c", snippet], env=env)


def main(argv=None):
    parser = argparse.ArgumentParser(
        description="Test dependency versions across multiple Python versions"
    )
    parser.add_argument(
        "--venv-root",
        default=str(Path.cwd() / ".venv-deps"),
        help="Directory to store created virtual environments",
    )
    parser.add_argument(
        "--samples",
        type=int,
        default=3,
        help="Versions to sample per dependency",
    )
    parser.add_argument(
        "--python-versions",
        nargs="+",
        default=list(DEFAULT_PYTHON_VERSIONS),
        metavar="VERSION",
        help=(
            "Python MAJOR.MINOR versions to test. Values may also be comma-separated. "
            f"Default: {', '.join(DEFAULT_PYTHON_VERSIONS)}"
        ),
    )
    parser.add_argument(
        "--no-optional",
        action="store_true",
        help="Skip optional dependency testing",
    )
    parser.add_argument(
        "--bounds-only",
        action="store_true",
        help="Run only the oldest and newest compatible dependency profiles",
    )
    args = parser.parse_args(argv)

    if args.samples < 2:
        parser.error("--samples must be at least 2 to exercise dependency diversity")

    try:
        python_versions = _normalize_python_versions(args.python_versions)
    except ValueError as exc:
        parser.error(str(exc))

    if not _uv_available():
        parser.error("uv is required; install the dev extras or install uv separately")

    package_root = Path(__file__).resolve().parents[1]
    venv_root = Path(args.venv_root)
    venv_root.mkdir(parents=True, exist_ok=True)

    succeeded = []
    failed_or_skipped = []

    for python_version in python_versions:
        python_root = venv_root / f"py-{python_version}"
        resolver_venv_path = python_root / "_resolver"

        print(f"\n######## Python {python_version} ########")
        _ensure_python(python_version)
        print(f"Setting up resolver venv: {resolver_venv_path}")
        _create_venv(resolver_venv_path, python_version)

        resolver_python = _venv_python(resolver_venv_path)
        _assert_python_version(resolver_python, python_version)

        print(f"Using resolver python: {resolver_python}")
        _run(
            [
                "uv",
                "pip",
                "install",
                "-p",
                resolver_python,
                "-U",
                "pip",
                "setuptools",
                "wheel",
                "packaging",
            ]
        )

        ranges = _dependency_ranges_for_python(python_version)
        core_versions = {
            name: _resolve_versions(resolver_python, name, min_v, max_v, args.samples)
            for name, (min_v, max_v) in ranges.items()
        }
        optional_versions = {
            name: _resolve_versions(resolver_python, name, min_v, max_v, args.samples)
            for name, (min_v, max_v) in OPTIONAL_RANGES.items()
        }

        scenarios = _build_scenarios(
            core_versions,
            optional_versions,
            not args.no_optional,
            bounds_only=args.bounds_only,
        )

        for scenario_name, scenario_deps in scenarios:
            scenario_label = f"py{python_version}:{scenario_name}"
            venv_path = python_root / scenario_name
            scenario_optional_ranges = OPTIONAL_RANGES if not args.no_optional else {}
            requirements, resolution = _requirements_for_scenario(
                scenario_name,
                scenario_deps,
                ranges,
                scenario_optional_ranges,
            )

            print(f"\n   ========= Scenario: {scenario_label} =========")

            try:
                print(f"Setting up venv: {venv_path}")
                _create_venv(venv_path, python_version)

                venv_python = _venv_python(venv_path)
                _assert_python_version(venv_python, python_version)

                print(f"Using venv python: {venv_python}")

                _install_dependencies(venv_python, requirements, resolution)
                _install_package(venv_python, package_root)
                _run_smoke_test(venv_python, include_optional=not args.no_optional)

                print(f"Scenario succeeded: {scenario_label}")
                succeeded.append(scenario_label)

            except subprocess.CalledProcessError as exc:
                print("Scenario failed; continuing to the next scenario.")
                failed_or_skipped.append(scenario_label)
                if exc.stdout:
                    print(exc.stdout)
                continue

            except (OSError, RuntimeError, ValueError) as exc:
                print(
                    "Scenario failed or was skipped; continuing to the next scenario."
                )
                failed_or_skipped.append(scenario_label)
                print(exc)
                continue

    print("\n   ========= Summary =========")
    print(f"Succeeded: {len(succeeded)}")
    print(f"Failed or skipped: {len(failed_or_skipped)}")

    if failed_or_skipped:
        print("❌ Failed or skipped scenarios:")
        for scenario_name in failed_or_skipped:
            print(f"  - {scenario_name}")
        print("Result: ❌ Some runs failed or were skipped.")
        return 1

    print("Result: ✅ All version runs succeeded.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: javascript/scripts/test-dependency-matrix.js
import { spawnSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

export const DEPENDENCY_RANGES = Object.freeze({
	'@huggingface/tokenizers': '^0.1.0',
	'onnxruntime-node': '^1.14.0',
	openai: '^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0',
});

const DEFAULT_SAMPLES = 3;
const scriptDirectory = path.dirname(fileURLToPath(import.meta.url));
const packageRoot = path.resolve(scriptDirectory, '..');

function run(
	command,
	args,
	{ cwd = packageRoot, env = process.env, inherit = false } = {},
) {
	const result = spawnSync(command, args, {
		cwd,
		env,
		encoding: 'utf8',
		stdio: inherit ? 'inherit' : 'pipe',
	});

	if (result.error) {
		throw result.error;
	}

	if (result.status !== 0) {
		const output = [result.stdout, result.stderr]
			.filter(Boolean)
			.join('\n')
			.trim();
		throw new Error(
			`${command} ${args.join(' ')} failed with exit code ${result.status ?? 1}${
				output ? `\n${output}` : ''
			}`,
		);
	}

	return (result.stdout ?? '').trim();
}

function runNpm(args, options = {}) {
	const command = process.platform === 'win32' ? 'npm.cmd' : 'npm';
	return run(command, args, options);
}

export function sampleVersions(versions, count) {
	if (versions.length === 0) {
		return [];
	}
	if (versions.length <= count) {
		return [...versions];
	}

	const selected = [];
	for (let index = 0; index < count; index += 1) {
		const versionIndex = Math.round(
			(index * (versions.length - 1)) / (count - 1),
		);
		const version = versions[versionIndex];
		if (!selected.includes(version)) {
			selected.push(version);
		}
	}
	return selected;
}

function parsePublishedVersions(output, dependencyName) {
	let parsed;
	try {
		parsed = JSON.parse(output);
	} catch (error) {
		throw new Error(
			`Unable to parse published versions for ${dependencyName}: ${error.message}`,
		);
	}

	const versions = Array.isArray(parsed) ? parsed : [parsed];
	const normalized = versions.filter(
		(version) => typeof version === 'string' && version.length > 0,
	);
	if (normalized.length === 0) {
		throw new Error(`No published versions found for ${dependencyName}`);
	}
	return normalized;
}

export function selectDependencyVersions(
	versions,
	latestVersion,
	samples,
	dependencyName = 'dependency',
) {
	if (!versions.includes(latestVersion)) {
		throw new Error(
			`${dependencyName}@latest (${latestVersion}) is outside the supported range`,
		);
	}

	const selected = sampleVersions(versions, samples);
	if (!selected.includes(latestVersion)) {
		selected[selected.length - 1] = latestVersion;
	}
	return [...new Set(selected)];
}

function resolveDependencyVersions(dependencyName, range, samples) {
	const output = runNpm([
		'view',
		`${dependencyName}@${range}`,
		'version',
		'--json',
	]);
	const latestOutput = runNpm(['view', dependencyName, 'version', '--json']);
	const latestVersion = parsePublishedVersions(
		latestOutput,
		dependencyName,
	)[0];
	return selectDependencyVersions(
		parsePublishedVersions(output, dependencyName),
		latestVersion,
		samples,
		dependencyName,
	);
}

export function buildScenarios(
	dependencyVersions,
	{ boundsOnly = false } = {},
) {
	const entries = Object.entries(dependencyVersions);
	const minimums = Object.fromEntries(
		entries.map(([name, versions]) => [name, versions[0]]),
	);
	const maximums = Object.fromEntries(
		entries.map(([name, versions]) => [name, versions.at(-1)]),
	);

	const scenarios = [
		{
			name: 'base-min',
			dependencies: minimums,
			expectedVersions: minimums,
		},
		{
			name: 'base-max',
			dependencies: maximums,
			expectedVersions: maximums,
		},
	];

	if (boundsOnly) {
		return scenarios;
	}

	for (const [targetName, versions] of entries) {
		for (const version of versions.slice(1, -1)) {
			const dependencies = {
				...minimums,
				[targetName]: version,
			};
			const slug = targetName.replace(/^@/, '').replaceAll('/', '-');
			scenarios.push({
				name: `dep-${slug}-${version}`,
				dependencies,
				expectedVersions: dependencies,
			});
		}
	}

	return scenarios;
}

export function collectDependencyVersions(tree, dependencyName) {
	const versions = [];
	const visit = (node) => {
		if (!node || typeof node !== 'object') {
			return;
		}
		const dependencies = node.dependencies;
		if (!dependencies || typeof dependencies !== 'object') {
			return;
		}

		const dependency = dependencies[dependencyName];
		if (dependency && typeof dependency.version === 'string') {
			versions.push(dependency.version);
		}
		for (const child of Object.values(dependencies)) {
			visit(child);
		}
	};

	visit(tree);
	return [...new Set(versions)];
}

function createPackageTarball(workRoot) {
	const output = runNpm(['pack', '--json', '--pack-destination', workRoot], {
		cwd: packageRoot,
	});
	const packed = JSON.parse(output);
	const filename = packed?.[0]?.filename;
	if (!filename) {
		throw new Error('npm pack did not return a package filename');
	}
	return path.resolve(workRoot, filename);
}

function createScenarioProject(scenarioDirectory, tarballPath, scenario) {
	fs.rmSync(scenarioDirectory, { recursive: true, force: true });
	fs.mkdirSync(scenarioDirectory, { recursive: true });

	const packageJson = {
		name: `aag-${scenario.name}`,
		private: true,
		type: 'module',
		dependencies: {
			'agent-action-guard': pathToFileURL(tarballPath).href,
			...scenario.dependencies,
		},
	};

	fs.writeFileSync(
		path.join(scenarioDirectory, 'package.json'),
		`${JSON.stringify(packageJson, null, 2)}\n`,
	);
}

function verifyInstalledVersions(scenarioDirectory, expectedVersions) {
	const names = Object.keys(DEPENDENCY_RANGES);
	const output = runNpm(['ls', ...names, '--all', '--json'], {
		cwd: scenarioDirectory,
	});
	const tree = JSON.parse(output);

	for (const [dependencyName, expectedVersion] of Object.entries(
		expectedVersions,
	)) {
		const installedVersions = collectDependencyVersions(
			tree,
			dependencyName,
		);
		if (installedVersions.length === 0) {
			throw new Error(`${dependencyName} was not installed`);
		}
		if (
			installedVersions.length !== 1 ||
			installedVersions[0] !== expectedVersion
		) {
			throw new Error(
				`${dependencyName}: expected only ${expectedVersion}, found ${installedVersions.join(', ')}`,
			);
		}
	}
}

function runSmokeTest(scenarioDirectory) {
	const snippet = `
		const guard = await import('agent-action-guard');
		const tokenizers = await import('@huggingface/tokenizers');
		const ortModule = await import('onnxruntime-node');
		const ort = ortModule.InferenceSession ? ortModule : ortModule.default;
		const openaiModule = await import('openai');

		if (typeof guard.isActionHarmful !== 'function') throw new Error('Action Guard import failed');
		if (typeof tokenizers.Tokenizer !== 'function') throw new Error('Tokenizer API missing');
		if (typeof ort?.InferenceSession?.create !== 'function') throw new Error('ONNX Runtime API missing');
		if (typeof openaiModule.default !== 'function') throw new Error('OpenAI default export missing');

		await guard.classifier.loadModel();
		const embeddingModel = new guard.EmbeddingModel({ modelName: 'matrix-smoke' });
		const client = await embeddingModel.getClient();
		if (typeof client.embeddings?.create !== 'function') throw new Error('OpenAI embeddings API missing');
	`;

	run(process.execPath, ['--input-type=module', '--eval', snippet], {
		cwd: scenarioDirectory,
	});
}

function parseArgs(argv) {
	const args = {
		boundsOnly: false,
		samples: DEFAULT_SAMPLES,
		workRoot: path.join(packageRoot, '.dependency-matrix'),
	};

	for (let index = 0; index < argv.length; index += 1) {
		const value = argv[index];
		if (value === '--bounds-only') {
			args.boundsOnly = true;
			continue;
		}
		if (value === '--samples') {
			args.samples = Number.parseInt(argv[index + 1], 10);
			index += 1;
			continue;
		}
		if (value === '--work-root') {
			args.workRoot = path.resolve(argv[index + 1]);
			index += 1;
			continue;
		}
		throw new Error(`Unknown argument: ${value}`);
	}

	if (!Number.isInteger(args.samples) || args.samples < 2) {
		throw new Error('--samples must be an integer of at least 2');
	}
	return args;
}

export function main(argv = process.argv.slice(2)) {
	let args;
	try {
		args = parseArgs(argv);
	} catch (error) {
		console.error(error.message);
		return 2;
	}

	fs.mkdirSync(args.workRoot, { recursive: true });
	console.log(`Node runtime: ${process.version}`);
	console.log(`Dependency samples per package: ${args.samples}`);

	let dependencyVersions;
	let tarballPath;
	try {
		dependencyVersions = Object.fromEntries(
			Object.entries(DEPENDENCY_RANGES).map(([name, range]) => {
				const versions = resolveDependencyVersions(
					name,
					range,
					args.samples,
				);
				console.log(`${name}: ${versions.join(', ')}`);
				return [name, versions];
			}),
		);
		tarballPath = createPackageTarball(args.workRoot);
	} catch (error) {
		console.error(`Matrix setup failed: ${error.message}`);
		return 1;
	}

	const scenarios = buildScenarios(dependencyVersions, {
		boundsOnly: args.boundsOnly,
	});
	const failed = [];

	for (const scenario of scenarios) {
		console.log(`\n========= Scenario: ${scenario.name} =========`);
		const scenarioDirectory = path.join(
			args.workRoot,
			'scenarios',
			scenario.name,
		);

		try {
			createScenarioProject(scenarioDirectory, tarballPath, scenario);
			console.log('Installing scenario dependencies');
			runNpm(
				['install', '--no-package-lock', '--no-audit', '--no-fund'],
				{
					cwd: scenarioDirectory,
					inherit: true,
				},
			);
			verifyInstalledVersions(
				scenarioDirectory,
				scenario.expectedVersions,
			);
			console.log('Running package smoke test');
			runSmokeTest(scenarioDirectory);
			console.log(`Scenario succeeded: ${scenario.name}`);
		} catch (error) {
			failed.push(scenario.name);
			console.error(`Scenario failed: ${scenario.name}`);
			console.error(error.message);
		}
	}

	console.log('\n========= Summary =========');
	console.log(`Succeeded: ${scenarios.length - failed.length}`);
	console.log(`Failed: ${failed.length}`);
	if (failed.length > 0) {
		for (const scenarioName of failed) {
			console.log(`  - ${scenarioName}`);
		}
		return 1;
	}

	console.log('Result: all dependency-version scenarios succeeded.');
	return 0;
}

const isMain =
	!process.env.NODE_TEST_CONTEXT &&
	process.argv[1] &&
	import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (isMain) {
	process.exitCode = main();
}
